feat: Add nemo customizer

raspawar opened the pull request:
# What does this PR do?

This PR adds support for NVIDIA's NeMo Customizer API to the Llama Stack post-training module. The integration enables users to fine-tune models using NVIDIA's cloud-based customization service through a consistent Llama Stack interface.


[//]: # (If resolving an issue, uncomment and update the line below)
[//]: # (Closes #[issue-number])

## Test Plan
[Describe the tests you ran to verify your changes with result summaries. *Provide clear instructions so the plan can be easily re-executed.*]
Yet to be done

Things pending under this PR:

- [x] Integration of fine-tuned model(new checkpoint) for inference with nvidia llm distribution
- [x] distribution integration of API
- [x] Add test cases for customizer(In Progress)
- [x] Documentation

```

LLAMA_STACK_BASE_URL=http://localhost:5002 pytest -v tests/client-sdk/post_training/test_supervised_fine_tuning.py 

============================================================================================================================================================================ test session starts =============================================================================================================================================================================
platform linux -- Python 3.10.0, pytest-8.3.4, pluggy-1.5.0 -- /home/ubuntu/llama-stack/.venv/bin/python
cachedir: .pytest_cache
metadata: {'Python': '3.10.0', 'Platform': 'Linux-6.8.0-1021-gcp-x86_64-with-glibc2.35', 'Packages': {'pytest': '8.3.4', 'pluggy': '1.5.0'}, 'Plugins': {'nbval': '0.11.0', 'metadata': '3.1.1', 'anyio': '4.8.0', 'html': '4.1.1', 'asyncio': '0.25.3'}}
rootdir: /home/ubuntu/llama-stack
configfile: pyproject.toml
plugins: nbval-0.11.0, metadata-3.1.1, anyio-4.8.0, html-4.1.1, asyncio-0.25.3
asyncio: mode=strict, asyncio_default_fixture_loop_scope=None
collected 2 items                                                                                                                                                                                                                                                                                                                                                            

tests/client-sdk/post_training/test_supervised_fine_tuning.py::test_post_training_provider_registration[txt=8B] PASSED                                                                                                                                                                                                                                                 [ 50%]
tests/client-sdk/post_training/test_supervised_fine_tuning.py::test_list_training_jobs[txt=8B] PASSED                                                                                                                                                                                                                                                                  [100%]

======================================================================================================================================================================== 2 passed, 1 warning in 0.10s ========================================================================================================================================================================
```
cc: @mattf @dglogo @sumitb

mattf commented on llama_stack/providers/remote/post_training/nvidia/utils.py:
```diff
@@ -0,0 +1,59 @@
+        RuntimeError: If the server is not running or ready
+    """
+    if not _is_nvidia_hosted(config):
+        logger.info("Checking NVIDIA NIM health...")
```
this should be for NeMo Customizer health instead

raspawar replied:
Yes will update this

SLR722 commented on llama_stack/providers/remote/post_training/nvidia/config.py:
```diff
@@ -0,0 +1,72 @@
+            "dataset_namespace": "${env.NVIDIA_DATASET_NAMESPACE:default}",
+            "access_policies": "${env.NVIDIA_ACCESS_POLICIES:}",
+            "project_id": "${env.NVIDIA_PROJECT_ID:test-project}",
+            "customizer_url": "${env.NVIDIA_CUSTOMIZER_URL:}",
```
provide default value for this?

SLR722 commented on llama_stack/providers/remote/post_training/nvidia/post_training.py:
```diff
@@ -0,0 +1,262 @@
+            logger_config: Dict[str, Any] - Configuration for logging
+        """
+        # map model to nvidia model name
+        model_mapping = {
```
shall we define this mapping outside this function? and do we also want to support llama3.2 3B model?

raspawar replied:
Yes will move this to the models.py and add supported models. This also reminds me to update the model name as per the llama-stack conventions: `CoreModelId.llama3_1_8b_instruct.value`

raspawar replied:
This is done @SLR722 ptal

SLR722 commented on llama_stack/providers/remote/post_training/nvidia/post_training.py:
```diff
@@ -0,0 +1,262 @@
+        job_config = {
+            "config": nvidia_model,
+            "dataset": {
+                "name": training_config["data_config"]["dataset_id"],
```
shall we use training_config.data_config.dataset_id instead?

SLR722 commented on llama_stack/providers/remote/post_training/nvidia/post_training.py:
```diff
@@ -0,0 +1,262 @@
+                "training_type": "sft",
+                "finetuning_type": "lora",
+                "epochs": training_config["n_epochs"],
+                "batch_size": training_config["data_config"]["batch_size"],
```
ditto

SLR722 commented on llama_stack/providers/remote/post_training/nvidia/post_training.py:
```diff
@@ -0,0 +1,262 @@
+            if isinstance(algorithm_config, dict) and algorithm_config.get("type") == "LoRA":
+                # Extract LoRA-specific parameters
+                lora_config = {k: v for k, v in algorithm_config.items() if k != "type"}
+                job_config["hyperparameters"]["lora"] = lora_config
```
seems hyperparameters.lora expects different from what lora_config has?

raspawar replied:
Yes adapter_dropout also needs support which I will add.
I am following up with nemo folks on llama-stack loraConfig parameters. Will keep you updated.

SLR722 commented on llama_stack/providers/remote/post_training/nvidia/post_training.py:
```diff
@@ -0,0 +1,262 @@
+                if training_config.get("algorithm_config", {}).get("adapter_dim"):
+                    job_config["hyperparameters"]["lora"]["adapter_dim"] = training_config["algorithm_config"][
+                        "adapter_dim"
+                    ]
```
nit: raise NotImplementedError() if not LoRA

SLR722 commented on llama_stack/providers/remote/post_training/nvidia/post_training.py:
```diff
@@ -0,0 +1,262 @@
+        page: Optional[int] = 1,
+        page_size: Optional[int] = 10,
+        sort: Optional[Literal["created_at", "-created_at"]] = "created_at",
+    ) -> ListNvidiaPostTrainingJobs:
```
Since you update the return type here, does it work with the lllama stack client SDK https://github.com/meta-llama/llama-stack-client-python/blob/9ce7520c75659e128dd7f6f8f83e5ea689ef2ebd/src/llama_stack_client/resources/post_training/job.py#L192?

SLR722 commented on llama_stack/providers/remote/post_training/nvidia/post_training.py:
```diff
@@ -0,0 +1,262 @@
+        return ListNvidiaPostTrainingJobs(data=jobs)
+
+    @webmethod(route="/post-training/job/status", method="GET")
+    async def get_training_job_status(self, job_uuid: str) -> Optional[NvidiaPostTrainingJob]:
```
ditto

SLR722 commented on tests/client-sdk/post_training/test_supervised_fine_tuning.py:
```diff
@@ -0,0 +1,386 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates.
```
shall we put the unit test under https://github.com/meta-llama/llama-stack/tree/main/tests/unit/providers?

SLR722 reviewed (commented):
thanks for the PR! add a few initial comments. Let me know if they make sense.

SLR722 commented:
Besides the unit tests in added (which is awesome!), could you also add E2E test in the test plan to actually call the nemo customizer endpoint and see if kick-off the training / get the finetuned checkpoint as expected? thanks!

raspawar commented on llama_stack/providers/remote/post_training/nvidia/config.py:
```diff
@@ -0,0 +1,72 @@
+        description="The NVIDIA API key, only needed of using the hosted service",
+    )
+
+    user_id: Optional[str] = Field(
```
So, the training has project, namespace related management with db integration. While llm inference does not, and user_id is something that is tagged for each customizer job for logging etc. 

@drazvan adding you in loop here, pls correct me/add relevant details.

mattf replied:
does this mean the caller gets to specify a string that gets logged?

is user_id string validated with the api_key?

can the user_id be omitted and the service determine it from the api_key?

raspawar replied:
does this mean the caller gets to specify a string that gets logged? ==> Yes, 
In get_jobs it can be passed in filter, to filter jobs started by a user_id

is user_id string validated with the api_key? ==> No atm, but it is tricky for the minikube deployment of nemo microservices(as we do not need API_KEY in that case).

raspawar replied:
I am going to skip this parameter for now

mattf commented on llama_stack/providers/remote/inference/nvidia/nvidia.py:
```diff
@@ -52,7 +52,7 @@
     convert_openai_completion_choice,
     convert_openai_completion_stream,
 )
-from .utils import _is_nvidia_hosted, check_health
```
is the post training inference endpoint guaranteed to have the same syntax and semantics of the nim inference endpoint?

raspawar commented:
Hi @SLR722
The linting/mypy was giving issues after I merged main

I have ignored this checks 
![Screenshot 2025-03-20 at 1 46 48 AM](https://github.com/user-attachments/assets/6b18bea6-0d14-4dd6-b403-d929f983b5e4)

SLR722 commented on llama_stack/distribution/ui/modules/api.py:
```diff
@@ -4,10 +4,12 @@
 from typing import Optional
 
-from llama_stack_client import LlamaStackClient
+from llama_stack_client import LlamaStackClient  # type: ignore[attr-defined]
```
why do we need this?

SLR722 commented on llama_stack/distribution/library_client.py:
```diff
@@ -52,6 +44,14 @@
     setup_logger,
     start_trace,
 )
+from llama_stack_client import (
```
can we avoid this kind of arbitrary format change?

SLR722 commented on llama_stack/distribution/ui/page/playground/rag.py:
```diff
@@ -5,12 +5,12 @@
 
 from llama_stack.distribution.ui.modules.api import llama_stack_api
 from llama_stack.distribution.ui.modules.utils import data_url_from_file
+from llama_stack_client.lib.agents.agent import Agent
```
ditto

SLR722 commented on llama_stack/providers/remote/inference/passthrough/passthrough.py:
```diff
@@ -29,6 +27,7 @@
 from llama_stack.apis.models import Model
 from llama_stack.distribution.library_client import convert_pydantic_to_json_value, convert_to_pydantic
 from llama_stack.providers.utils.inference.model_registry import ModelRegistryHelper
+from llama_stack_client import AsyncLlamaStackClient
```
ditto

SLR722 reviewed (commented):
thanks! I saw several auto format thing that actually did nothing. Could you avoid doing that by disabling the auto formatting in you IDE?

SLR722 commented:
> Hi @SLR722 The linting/mypy was giving issues after I merged main
> 
> I have ignored this checks ![Screenshot 2025-03-20 at 1 46 48 AM](https://private-user-images.githubusercontent.com/168514198/424689456-6b18bea6-0d14-4dd6-b403-d929f983b5e4.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDI0MjgzNTEsIm5iZiI6MTc0MjQyODA1MSwicGF0aCI6Ii8xNjg1MTQxOTgvNDI0Njg5NDU2LTZiMThiZWE2LTBkMTQtNGRkNi1iNDAzLWQ5MjlmOTgzYjVlNC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwMzE5JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDMxOVQyMzQ3MzFaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1hMWU2YmY4MjZmYWEyZmQ4MmM5OTY0Y2VmNjFjM2IzNzdiYmYxOTgyMjgzNDI0MDMwMzg0YjA2MzAzOTZjNDc4JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.ozZZsg5vImBJlMLaoJDjUGYi4hotkzMMK76-eMR07ik)
> 
> ptal: looks stable to me

where did you see this issue? on your IDE or somewhere else?

raspawar commented:
@SLR722, I have reverted the auto-correct linting changes caused by pre-commit.
Also disabled the ruff auto sorting by removing 
```toml
[tool.ruff.lint]
select = [
"I",   # isort, removed this
]
```

any thoughts?

SLR722 commented:
> > > @SLR722, I have reverted the auto-correct linting changes caused by pre-commit. Also disabled the ruff auto sorting by removing
> > > ```toml
> > > [tool.ruff.lint]
> > > select = [
> > > "I",   # isort, removed this
> > > ]
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > any thoughts?
> > 
> > 
> > oh, thanks for checking!. If the auto format is caused by pre-commit, then let's keep that. Let's just avoid the auto formatting by the IDE
> 
> I won't recommend that, as it is updating the changes out of scope. Also I do not understand why unit test cases are failing as I see opentelemetry part of test dependency and it works locally. @SLR722 will you be able to check this?

It's because you are trying to initialize 'nvidia' template here https://github.com/meta-llama/llama-stack/pull/1448/files#diff-1897f3ec6f6486a0f22dfe59ba3f7784e0c2fa7f65cc8df4d695f3c7e1e579f6R31 and the template needs opentelemetry as dependency

In your local, you probably already installed opentelemetry in your conda env. So ii wouldn't raise error

SLR722 commented:
> Hey @SLR722,
> 
> I have removed the llamastackclient from unit tests and checking the `NvidiaPostTrainingAdapter` functionality alone. ptal!

thanks! could you link the integration test results which actually call the nemo endpoint in the test plan?

raspawar commented:
Integration test case results are already added in PR description.

I already resolved merge conflicts thrice and I see distribution is not getting up because safety adpater is not getting built.
Can we merge this first? It is stale for **3 weeks now**

SLR722 reviewed (approved):
Thank you for your patience in addressing all the comments!

booxter commented:
@raspawar do you by chance know where I can find REST API definition for NeMo Customizer API? I'm working on Job API revamp here https://github.com/meta-llama/llama-stack/pull/1772 and I need to know how the provider API behaves so that I can adjust the provider to work with the new API schema.

raspawar commented:
> @raspawar do you by chance know where I can find REST API definition for NeMo Customizer API? I'm working on Job API revamp here #1772 and I need to know how the provider API behaves so that I can adjust the provider to work with the new API schema.

Latest Nemo MS documentation: https://docs.nvidia.com/nim/large-language-models/latest/getting-started.html
